### ISSUE TITLE
Will `with operators from X` prevent polymorphism?

### ISSUE THREAD

**Issue by miyaokamarina** [littledan/proposal-operator-overloading#14]
E.g. we have `Vec` class with `*` operator overloaded like

```typescript
function dotProduct(l: Vec, r: Vec) {
  return l.a * r.a + l.b * r.b + l.c * r.c;
}
```

Then, if we will define `Px`, `Em`, etc., etc. classes with overloaded `*` and `+`, it _seems_ possible to create `Vec` of some of these classes, but, if I understand correctly, currently `with operators from X` for _each_ of these classes is required in `dotProduct`.

**Comment by littledan** [comment]
Yes, I think yo uunderstood correctly. The idea of this proposal is that you'd use `with operators from` on each of the classes, and then you have the polymorphism enabled within that lexical scope. Or, you could define a single superclass that dispatches internally off of its sublclasses.

**Comment by miyaokamarina** [comment]
Why we need to specify each class to use operators from explicitly? If we just need to prevent behavior of existing code to be changed by passing a value with overloaded operators, maybe something like `'use operators';` will be more suitable? Lexical scopes w/o this directive will work as expected, but scopes with it are assumed to know that values’ `[[OperatorsSet]]`s may be overloaded. So the behavior of existing code remains unchanged, but operators-aware code is… operators-aware.

(IIRC, there was some proposal on directives/pragmas in general? Is it still active?)

**Comment by littledan** [comment]
This proposal attempts to get at some strong [predictability goals](https://github.com/littledan/proposal-operator-overloading#design-goals) based on my understanding of some committee member's requests (cc @BrendanEich). It's possible that these goals could be tweaked if we think it through more.

The concern here was about "injection" of unexpected behavior through passing an unexpected class into a site that has operator overloading. I interpreted that to mean that, even if you expected `+` to work a certain way on Vectors and Matrices, you might not be thinking about some other class, so it would still constitute this sort of injection risk.

There's another pragma proposed at https://github.com/tc39/proposal-function-implementation-hiding . I don't use a string pragma here because (1) I really don't like string pragmas, and have expressed this opinion in the context of that other proposal (2) a string pragma could not refer to values this way.

**Comment by miyaokamarina** [comment]
But injection risk exists even in the case of

> a single superclass that dispatches internally off of its subclasses.

I can create `NumberLike` class with `*` and `+` overloaded, then use operators from it in `dotProduct`, then create a subclass that e.g. has effects in `+` implementation; or even returns a value of wrong type. `dotProduct` will never expect such behavior, and this risk cannot be prevented using current design (probably it can’t be prevented entirely, even using TypeScript).

**Comment by littledan** [comment]
The theory would be that, in that case, you're opting in to trusting the superclass to do that, and not trusting other pieces of code that overload different operators that you don't know about.

I think we could use all sorts of different syntaxes for pragmas, but the unique thing about string pragmas is that they are backwards-compatible, in the sense that they are ignored in older JS implementations. This property is not very useful for operator overloading, but may be useful for function implementation hiding.

**Comment by miyaokamarina** [comment]
> The theory would be that, in that case, you're opting in to trusting the superclass to do that, and not trusting other pieces of code that overload different operators that you don't know about.

Isn’t this the TypeScript’s domain? It looks like introducing some kind of static typing into entirely dynamic JS.

**Comment by littledan** [comment]
I don't see how this relates to TypeScript. This is about solid runtime semantics, which TypeScript doesn't interfere with. Along the same lines, we're adding strong encapsulation with `#private` fields. The semantics here are, the object has to have the internal slot with the particular value, at runtime.

**Comment by justinfagnani** [comment]
It would really unfortunate to not be able to lose polymorphism with the opt-in approach here.

What is the purpose of restricting overloads to a fixed set of operator collections in the source? Once you opt-in to any overloads you'll have to have more hidden class / type checks, as I understand.

Without polymorphism, how would generic code like `sum()` or `mean()` be written?

**Comment by Jamesernator** [comment]
> Without polymorphism, how would generic code like sum() or mean() be written?

It does seem kinda unfortunate that these are unavailable. I wonder if a different solution would just be allow overloading operators in a specific context with a specific bag of operators.

e.g. :

```js
const PointOperators = Operators({ 
  add: (a, b) => ({ x: a.x + b.x, y: a.y + b.y })
});

with operators (PointOperators) {
  const a = { x: 10, y: 20 };
  const b = { x: 30, y: 40 };

  console.log(a + b); // { x: 40, y: 60 }
}
```

This would effectively just desugar into:

```js
console.log(PointOperators.add(a, b));
```

Although with such a scheme would mean it'd be necessary to also pass the operator provider but with decorators one might just use `.constructor`. e.g.:

```ts
type Summable = {
    constructor: Operators<{
        add(a: this, b: this): this;
    }
};

function sum<T extends Summable>(items: T[]) {
    with operators (a.constructor) {
        let total: A;
        for (const item of items) {
            total += item;
        }
        return total;
    }
}
```

It also means you could apply operator-overloading to existing types e.g.:

```ts
function galoisField(n: bigint) {
  if (!isPrimePower(n)) {
    throw new Error("n is not prime power");
  }

  return Operators({
    add(a: bigint, b: bigint): bigint {
      return (a + b) % n;
    },

    mul(a: bigint, b: bigint): bigint {
      return (a * b) % n;
    },

    negate(a: bigint) {
      return n - a;
    },

    inverse(a: bigint) {
      return fastPow(a, n-2, n);
    },

    div(a: bigint, b: bigint): bigint {
      if (b === 0) {
        throw new Error("Division by zero");
      }
      return a * fastPow(b, n-2, n);
    }
  });
}

with operators (galoisField(7n)) {
  const a = 4n;
  const b = 5n;

  console.log(a + b); // 2n
  console.log(a * b); // 6n
  console.log(a / b); // 5n
  console.log(-a);    // 3n
}
```

**Comment by littledan** [comment]
This proposal does permit polymorphism; it's just bound by the enclosing lexical declarations. It's possible to opt into operators on several classes at once.

The README explains the motivation for these restrictions in terms of meeting reliability guarantees. It seemed to me like several TC39 delegates viewed these guarantees positively when I presented operator overloading in the December 2019 TC39 meeting.

Maybe we could have a special "opt into everything" declaration, but I still feel like the normal usage should be to opt into specifically what you use. I would like code to be able to use some amount of operator overloading while still benefiting from the predictability benefits.

**Comment by miyaokamarina** [comment]
Why not to go further, and require explicit declaration of specific operators to use?

E.g., my code may expect `*` and `**` overloaded in type `T`, and accepts arguments of both `T` and `string`. And it will expect its string arguments to behave exactly as strings, i.e., `strArg1 + strArg2` should not produce anything but string. _(Welp, it still may produce a number, if arguments have `valueOf(): number;` or `[Symbol.toPrimitive](): number;` method.)_

In such case, the syntax may be:

```
operators * from T;         // Use all operators from `T`.
operators { *, ** } from T; // Use only `*` and `**` from `T`.

operators { *, ** } from *; // Use `*` and `**` from any type.
operators * from *;         // Use all operators from any type.
```

Also, this syntax looks like `export ... from ...` and may be accepted easier by developers.

**Comment by littledan** [comment]
Yeah, we could do that. I'd thought about that possibility vaguely but I didn't have much of a syntax in mind--I like your suggestion, if we want to go with that. We might want to tweak it slightly just so parsing doesn't require three tokens of lookahead, though.

I think this choice is something of a spectrum that we can make a tradeoff within, between detailed and general. There are lots of possibilities. There are also some costs to making the language quite so rich. I think we can figure out what tradeoff to make if we consider a lot of practical usage scenarios.

**Comment by Jamesernator** [comment]
Just for the record is the RHS of these forms going to be an expression or an identifier?

e.g. Is `<some operators syntax from> point.constructor;` allowed?

**Comment by littledan** [comment]
@Jamesernator I was imagining an arbitrary expression.

**Comment by Jamesernator** [comment]
Cool cause that means things like `sum()` and stuff can still be written without needing extra arguments (e.g. without `sum(points, Point)`)

**Comment by justinfagnani** [comment]
@littledan 
> I still feel like the normal usage should be to opt into specifically what you use

This is something I don't quite get, regarding the intended usage. How would generic libraries know what types to opt into? It seems like every function would have to be written specifically for one set of types. Will we see `meanPrimitives()`, `meanVec()`, `meanMoney()`, `meanComplex()`, etc...?

**Comment by miyaokamarina** [comment]
@justinfagnani, maybe something like this:

```javascript
const meanA = (Class, ...xs) => {
  operators { +, / } from Class;

  return xs.reduce((a, b) => a + b, Class.zero()) / xs.length;
};

const meanB = (...xs) => {
  operators { +, / } from *;

  return xs.reduce((a, b) => a + b, 0) / xs.length;
};
```

**Comment by littledan** [comment]
Could anyone give an example of these generic libraries, a little more concretely than names of some functions? What would they be used for?

**Comment by miyaokamarina** [comment]
E.g., for generic `Math` implementation:

```javascript
export const min = (x, ...ys) => {
  operators { < } from *;

  for (conts y of ys) {
    if (y < x) {
      x = y;
    }
  }

  return x;
};

min(3@px, 4@px); // 3@px; CSS units
min(5@mb, 6@kb); // 6@kb; custom units
min(Vec2d(3, 4), Vec2d(5, 6)); // Vec2d(3, 4); custom type with `<`

min(fantasyLandOrd1, fantasyLandOrd2); // ...; Fantasy Land `Ord` typeclass
```

**Comment by justinfagnani** [comment]
@littledan Lodash has a few math functions, like `min()`: https://lodash.com/docs/4.17.15#min

SimpleStatistics has a whole bunch of functions that could be generic: https://simplestatistics.org/docs/

There are a bunch of other search hits, but honestly it looks like a lot of numeric libraries from 4-7 years ago have bitrotted.

**Comment by littledan** [comment]
What I don't understand is, when you want to use `min` on custom-defined types, as opposed to on one specific type. I don't want to nit-pick your examples too much, but I don't think `<` makes much sense on CSS units (without more context), and on vectors I'm not sure what it returns (another vector? or is it comparing magnitudes?).

My intuition is that, for things like `Math.min`, if we decided to extend it for additional built-in numeric types, it'd still just be for those, and not for arbitrary user-defined values. You could still faithfully polyfill new numeric types by redefining `Math.min` to include all the types. This way, `Math.min` doesn't call into unanticipated user code (a core goal of operator overloading).
